# Определение перспективного тарифа для телеком-компании

Цели и исследования:

- Определить какой тарифный план приносит наибольший доход для компании

Так же необходимо ответить на следующие вопросы:

- Сколько звонков делает каждый пользователь в месяц
- Сколько тратит минут каждый пользователь в месяц
- Сколько сообщений отправляет каждый пользователь в месяц
- Какой объем интернет трафика использует каждый пользователь в месяц
- Какая выручка с каждого пользователя в месяц
- Определить сколько минут, сообщений и трафика требуется пользователям для каждого тарифа в месяц

Проверить следующие гипотезы:

- Средняя выручка пользователей между тарифами «Ультра» и «Смарт» различаются
- Средняя выручка пользователей в Москве отличаются от средней выручки пользователей из других регионов


## Импорт библиотек и данных

In [ ]:
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt

from datetime import datetime, timedelta

In [2]:
calls = pd.read_csv('data/raw/calls.csv')
internet = pd.read_csv('data/raw/internet.csv')
messages = pd.read_csv('data/raw/messages.csv')
tariffs = pd.read_csv('data/raw/tariffs.csv')
users = pd.read_csv('data/raw/users.csv')

### Описание полей таблиц

**Таблица users**
- `user_id` — уникальный идентификатор пользователя
- `first_name` — имя пользователя
- `last_name` — фамилия пользователя
- `age` — возраст пользователя 
- `reg_date` — дата подключения тарифа (день, месяц, год)
- `churn_date` — дата прекращения пользования тарифом 
- `city` — город проживания пользователя
- `tariff` — название тарифного плана

**Таблица calls**
- `id` — уникальный номер звонка
- `call_date` — дата звонка
- `duration` — длительность звонка в минутах
- `user_id` — идентификатор пользователя, сделавшего звонок

**Таблица messages**
- `id` — уникальный номер сообщения
- `message_date` — дата сообщения
- `user_id` — идентификатор пользователя, отправившего сообщение

**Таблица internet**
- `id` — уникальный номер сессии
- `mb_used` — объём потраченного за сессию интернет-трафика (в мегабайтах)
- `session_date` — дата интернет-сессии
- `user_id` — идентификатор пользователя

**Таблица tariffs**
- `tariff_name` — название тарифа
- `rub_monthly_fee` — ежемесячная абонентская плата в рублях
- `minutes_included` — количество минут разговора в месяц, включённых в абонентскую плату
- `messages_included` — количество сообщений в месяц, включённых в абонентскую плату
- `mb_per_month_included` — объём интернет-трафика, включённого в абонентскую плату (в мегабайтах)
- `rub_per_minute` — стоимость минуты разговора сверх тарифного пакета (например, если в тарифе 100 минут разговора в месяц, то со 101 минуты будет взиматься плата)
- `rub_per_message` — стоимость отправки сообщения сверх тарифного пакета
- `rub_per_gb` — стоимость дополнительного гигабайта интернет-трафика сверх тарифного пакета (1 гигабайт = 1024 мегабайта)

## Знакомство с данными

In [3]:
def get_info(df):
  print(f"Кол-во строк: {len(df)} \nКол-во колонок: {df.shape[1]}\n"
    f"Дубликаты: {df.duplicated().sum()}\n"
    f"Память: {df.memory_usage(deep=True).sum() / 1e6:.1f} MB")

  display(df.head())
  print('___________')
  
  print(f"{'GENERAL'}")
  profile = pd.DataFrame({
    "dtype": df.dtypes.astype(str),
    "nunique": df.nunique(dropna=False),
    "na": df.isna().sum(),
    "na%": (df.isna().mean() * 100).round(2),
  })
  display(profile.T)
  
  print('___________')
  num = df.select_dtypes(include="number")
  if not num.empty:
    print(f"DESCRIBE")
    display(num.describe(percentiles=[0.01, 0.05, 0.25, 0.5, 0.75, 0.95, 0.99]).round(2))

In [4]:
HEADER = "\033[0;37;40m"
RESET = "\033[0m"
tables = [calls, internet, messages, tariffs, users]
names = ['calls', 'internet', 'messages', 'tariffs', 'users']

for table, name in zip(tables, names):
  print(f'{HEADER} ___{name.upper()}___ {RESET}')
  get_info(table)
  print(f'Конец \n')


 ___CALLS___ 
Кол-во строк: 202607 
Кол-во колонок: 4
Дубликаты: 0
Память: 26.7 MB


,id,call_date,duration,user_id
0,1000_0,2018-07-25,0.00,1000
1,1000_1,2018-08-17,0.00,1000
2,1000_2,2018-06-11,2.85,1000
3,1000_3,2018-09-21,13.80,1000
4,1000_4,2018-12-15,5.18,1000


___________
GENERAL


,id,call_date,duration,user_id
dtype,str,str,float64,int64
nunique,202607,365,2871,492
na,0,0,0,0
na%,0.0,0.0,0.0,0.0


___________
DESCRIBE


,duration,user_id
count,202607.00,202607.00
mean,6.76,1253.94
std,5.84,144.72
min,0.00,1000.00
1%,0.00,1005.00
5%,0.00,1027.00
25%,1.30,1126.00
50%,6.00,1260.00
75%,10.70,1379.00
95%,17.52,1472.00


Конец 

 ___INTERNET___ 
Кол-во строк: 149396 
Кол-во колонок: 5
Дубликаты: 0
Память: 20.9 MB


,Unnamed: 0,id,mb_used,session_date,user_id
0,0,1000_0,112.95,2018-11-25,1000
1,1,1000_1,1052.81,2018-09-07,1000
2,2,1000_2,1197.26,2018-06-25,1000
3,3,1000_3,550.27,2018-08-22,1000
4,4,1000_4,302.56,2018-09-24,1000


___________
GENERAL


,Unnamed: 0,id,mb_used,session_date,user_id
dtype,int64,str,float64,str,int64
nunique,149396,149396,70003,365,497
na,0,0,0,0,0
na%,0.0,0.0,0.0,0.0,0.0


___________
DESCRIBE


,Unnamed: 0,mb_used,user_id
count,149396.00,149396.00,149396.00
mean,74697.50,370.19,1252.10
std,43127.05,278.30,144.05
min,0.00,0.00,1000.00
1%,1493.95,0.00,1006.00
5%,7469.75,0.00,1025.00
25%,37348.75,138.19,1130.00
50%,74697.50,348.02,1251.00
75%,112046.25,559.55,1380.00
95%,141925.25,866.52,1476.00


Конец 

 ___MESSAGES___ 
Кол-во строк: 123036 
Кол-во колонок: 3
Дубликаты: 0
Память: 15.2 MB


,id,message_date,user_id
0,1000_0,2018-06-27,1000
1,1000_1,2018-10-08,1000
2,1000_2,2018-08-04,1000
3,1000_3,2018-06-16,1000
4,1000_4,2018-12-05,1000


___________
GENERAL


,id,message_date,user_id
dtype,str,str,int64
nunique,123036,364,426
na,0,0,0
na%,0.0,0.0,0.0


___________
DESCRIBE


,user_id
count,123036.00
mean,1256.99
std,143.52
min,1000.00
1%,1004.00
5%,1026.00
25%,1134.00
50%,1271.00
75%,1381.00
95%,1475.00


Конец 

 ___TARIFFS___ 
Кол-во строк: 2 
Кол-во колонок: 8
Дубликаты: 0
Память: 0.0 MB


,messages_included,mb_per_month_included,minutes_included,rub_monthly_fee,rub_per_gb,rub_per_message,rub_per_minute,tariff_name
0,50,15360,500,550,200,3,3,smart
1,1000,30720,3000,1950,150,1,1,ultra


___________
GENERAL


,messages_included,mb_per_month_included,minutes_included,rub_monthly_fee,rub_per_gb,rub_per_message,rub_per_minute,tariff_name
dtype,int64,int64,int64,int64,int64,int64,int64,str
nunique,2,2,2,2,2,2,2,2
na,0,0,0,0,0,0,0,0
na%,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0


___________
DESCRIBE


,messages_included,mb_per_month_included,minutes_included,rub_monthly_fee,rub_per_gb,rub_per_message,rub_per_minute
count,2.00,2.00,2.00,2.00,2.00,2.00,2.00
mean,525.00,23040.00,1750.00,1250.00,175.00,2.00,2.00
std,671.75,10861.16,1767.77,989.95,35.36,1.41,1.41
min,50.00,15360.00,500.00,550.00,150.00,1.00,1.00
1%,59.50,15513.60,525.00,564.00,150.50,1.02,1.02
5%,97.50,16128.00,625.00,620.00,152.50,1.10,1.10
25%,287.50,19200.00,1125.00,900.00,162.50,1.50,1.50
50%,525.00,23040.00,1750.00,1250.00,175.00,2.00,2.00
75%,762.50,26880.00,2375.00,1600.00,187.50,2.50,2.50
95%,952.50,29952.00,2875.00,1880.00,197.50,2.90,2.90


Конец 

 ___USERS___ 
Кол-во строк: 500 
Кол-во колонок: 8
Дубликаты: 0
Память: 0.2 MB


,user_id,age,churn_date,city,first_name,last_name,reg_date,tariff
0,1000,52,NaN,Краснодар,Рафаил,Верещагин,2018-05-25,ultra
1,1001,41,NaN,Москва,Иван,Ежов,2018-11-01,smart
2,1002,59,NaN,Стерлитамак,Евгений,Абрамович,2018-06-17,smart
3,1003,23,NaN,Москва,Белла,Белякова,2018-08-17,ultra
4,1004,68,NaN,Новокузнецк,Татьяна,Авдеенко,2018-05-14,ultra


___________
GENERAL


,user_id,age,churn_date,city,first_name,last_name,reg_date,tariff
dtype,int64,int64,str,str,str,str,str,str
nunique,500,58,34,76,246,419,270,2
na,0,0,462,0,0,0,0,0
na%,0.0,0.0,92.4,0.0,0.0,0.0,0.0,0.0


___________
DESCRIBE


,user_id,age
count,500.00,500.00
mean,1249.50,46.59
std,144.48,16.67
min,1000.00,18.00
1%,1004.99,18.00
5%,1024.95,21.00
25%,1124.75,32.00
50%,1249.50,46.00
75%,1374.25,62.00
95%,1474.05,72.05


Конец 



### Краткий вывод по данным

**Таблица `calls`**
- В таблице 202607 записи, дубликатов и пропусков нет. 
- Все IDs уникальные. 
- Колонка `call_date` имеет тип данных `str` в следующем шаге необходимо преобразовать ее в `datetime`
- `duration` до 5го процентиля равен 0 - это недозвоны

**Таблица `internet`**
- В таблице 149396 записей, дубликатов и пропусков нет. 
- `session_date` имеет тип данных `str` в следующем шаге необходимо преобразовать ее в `datetime`
- Лишняя колонка `unnamed` дублирующая индекс

**Таблица `messages`**
- В таблице 123036 записей, дубликатов и пропусков нет. 
- `message_date` имеет тип данных `str` в следующем шаге необходимо преобразовать ее в `datetime`

**Таблица `tariffs`**
- 2 строки с описанием планов
- Нет недостающих значений, нет дубликатов, все данные соответствуют типу данных

**Таблица `users`**
- В таблице 500 записей, дубликатов. 
- `reg_date`, `churn_date` имеют тип данных `str` в следующем шаге необходимо преобразовать ее в `datetime`
- В колонке `churn_date` есть пропущенные значения, свзязанно это с тем, что у пользователя еще действует тариф
